balancer-js/src/modules/pools/index.ts

```diff
@@ -0,0 +1,205 @@
+import type {
+  BalancerNetworkConfig,
+  BalancerDataRepositories,
+  Findable,
+  Searchable,
+  Pool,
+  PoolWithMethods,
+  AprBreakdown,
+  PoolAttribute,
+} from '@/types';
+import { JoinPoolAttributes } from './pool-types/concerns/types';
+import { PoolTypeConcerns } from './pool-type-concerns';
+import { PoolApr } from './apr/apr';
+import { Liquidity } from '../liquidity/liquidity.module';
+import { Join } from '../joins/joins.module';
+
+/**
+ * Controller / use-case layer for interacting with pools data.
+ */
+export class Pools implements Findable<PoolWithMethods> {
+  aprService;
+  liquidityService;
+  joinService;
+
+  constructor(
+    private networkConfig: BalancerNetworkConfig,
+    private repositories: BalancerDataRepositories
+  ) {
+    this.aprService = new PoolApr(
+      this.repositories.pools,
+      this.repositories.tokenPrices,
+      this.repositories.tokenMeta,
+      this.repositories.tokenYields,
+      this.repositories.feeCollector,
+      this.repositories.yesterdaysPools,
+      this.repositories.liquidityGauges,
+      this.repositories.feeDistributor
+    );
+    this.liquidityService = new Liquidity(
+      repositories.pools,
+      repositories.tokenPrices
+    );
+    this.joinService = new Join(this.repositories.pools, networkConfig.chainId);
+  }
+
+  dataSource(): Findable<Pool, PoolAttribute> & Searchable<Pool> {
+    // TODO: Add API data repository to data and use liveModelProvider as fallback
+    return this.repositories.pools;
+  }
+
+  /**
+   * Calculates APR on any pool data
+   *
+   * @param pool
+   * @returns
+   */
+  async apr(pool: Pool): Promise<AprBreakdown> {
+    return this.aprService.apr(pool);
+  }
+
+  /**
+   * Calculates total liquidity of the pool
+   *
+   * @param pool
+   * @returns
+   */
+  async liquidity(pool: Pool): Promise<string> {
+    return this.liquidityService.getLiquidity(pool);
+  }
+
+  /**
+   * Builds generalised join transaction
+   *
+   * @param poolId          Pool id
+   * @param expectedBPTOut  Minimum BPT expected out of the join transaction
+   * @param tokens          Token addresses
+   * @param amounts         Token amounts in EVM scale
+   * @param userAddress     User address
+   * @param wrapMainTokens  Indicates whether main tokens should be wrapped before being used
+   * @param authorisation   Optional auhtorisation call to be added to the chained transaction
+   * @returns transaction data ready to be sent to the network
+   */
+  async generalisedJoin(
+    poolId: string,
+    expectedBPTOut: string,
+    tokens: string[],
+    amounts: string[],
+    userAddress: string,
+    wrapMainTokens: boolean,
+    authorisation?: string
+  ): Promise<{
+    to: string;
+    data: string;
+    decode: (output: string) => string;
+  }> {
+    return this.joinService.joinPool(
```

Comment:
@johngrantuk - this is the only relevant change where `generalisedJoin` was removed from within the `wrap` method.
I believe it makes more sense to structure it like liquidity and apr implementations because it depends on having access to the pools repository and it does not depend on poolTypes as the other methods do.